pkg/registrymirror/registrymirrorsstore.go

```diff
@@ -0,0 +1,325 @@
+	s.timerMutex.Lock()
+	defer s.timerMutex.Unlock()
+
+	if s.timer != nil {
+		s.timer.Stop()
+	}
+
+	s.timer = time.AfterFunc(s.updateDelay, func() {
+		if err := s.updateConfigNow(); err != nil {
+			log.Errorf("Failed to update the registry mirror config: %v", err)
+		}
+	})
+
+	return nil
+}
+
+func (s *FileStore) updateConfigNow() error {
+	icspRules, idmsRules, itmsRules := s.getAllMirrorSets()
+
+	// Populate config.
+	config := new(sysregistriesv2.V2RegistriesConf)
+	err := registries.EditRegistriesConfig(config, nil, nil, icspRules, idmsRules, itmsRules)
+	if err != nil {
+		return errors.Wrap(err, "could not create registries config")
+	}
+
+	// Encode config.
+	var newData bytes.Buffer
+	encoder := toml.NewEncoder(&newData)
+	if err := encoder.Encode(config); err != nil {
+		return errors.Wrap(err, "could not encode registries config")
+	}
+
+	// Ensure output dir exists.
+	err = os.MkdirAll(filepath.Dir(s.configPath), 0755)
+	if err != nil {
+		return errors.Wrap(err, "could not create directories")
+	}
+
+	// Open handle to output file.
+	f, err := os.Create(s.configPath)
+	if err != nil {
+		return errors.Wrap(err, "could not create/open file")
+	}
+	defer utils.IgnoreError(f.Close)
+
+	// Write encoded config to output file.
+	_, err = f.Write(newData.Bytes())
+	if err != nil {
+		return errors.Wrap(err, "could not write bytes to file")
+	}
+
+	// Close file before defer to ensure close errors are surfaced.
+	if err := f.Close(); err != nil {
+		return errors.Wrap(err, "could not close file")
+	}
+
+	// Invalidate the registries cache so that the updated file will be read on next invocation of PullSources.
+	sysregistriesv2.InvalidateCache()
+
+	log.Debugf("Successfully updated the registries config at %q with %v icspRules, %v idmsRules, %v itmsRules",
+		s.configPath, len(icspRules), len(idmsRules), len(itmsRules))
+
+	return nil
+}
+
+// PullSources will return image references in the order they should be attempted when pulling the image.
+// The returned slice may (or may not) include the source image based on the registries config.
+// Source image must be fully qualified (include the registry hostname).
+func (s *FileStore) PullSources(srcImage string) ([]string, error) {
+	// FindRegistry will parse the registries config file and cache it for future usage.
+	reg, err := sysregistriesv2.FindRegistry(s.systemContext, srcImage)
+	// Using errors.Is instead of os.IsNotExist because the return from FindRegistry
+	// wraps the error in a type that os.IsNotExist ignores.
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, errors.Wrap(err, "failed finding registry")
+	}
+
+	if reg == nil {
+		return []string{srcImage}, nil
+	}
+
+	// ParseNamed assumes srcImage is a fully qualified references (contains a hostname), will produce error
+	// otherwise. In the future if support for short names is needed, use reference.ParseNormalizedNamed instead.
+	ref, err := reference.ParseNamed(srcImage)
```

Comment:
This isn't necessarily something specific to this PR, but more a generic curiosity: AFAIK, we currently assume `docker.io` as [the default hostname in case a short name is given within the an image reference](https://github.com/stackrox/stackrox/blob/master/pkg/images/utils/utils.go#L19).

I'm not sure what the "recommendation" by OpenShift is in ICSPs for hostname referring to "Docker Hub" (also presumably CRI-O as our container runtime on OpenShift). Just need to make sure this is somehow in-line to avoid not mirroring due to mirror config being for `registry-1.docker.io/<something>` but internally our images get assigned `docker.io/<something>`.

Does this apply / make sense or is it nothing we necessarily have to worry about?

Reply:
It does make sense and will eventually apply (see ROX-18689). Right now sensor will [use the 'registryOverride'](https://github.com/stackrox/stackrox/blob/794cff01e2eea62f17e18023708ace71d40f5ce1/sensor/kubernetes/listener/resources/convert.go#L353) from the clusters 'dynamic config' to qualify references that have no hostname - and if that is empty will use `docker.io` as you mentioned.

This only allows for a single override, whereas CRI-O & `c/image` allow for multiple short names to be specified and will use the first one that triggers a successful pull.